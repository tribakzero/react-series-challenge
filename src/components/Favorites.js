import React, { Component } from 'react';
import GifList from './GifList'

class Favorites extends Component {
  constructor(props) {
    super(props)
    this.state = {
      gifs: [
        {
          images: {
            original: {
              url: "https://media.giphy.com/media/l1J9KqKcpnG6ychfq/giphy.gif"
            }
          },
          liked: true
        },
        {
          images: {
            original: {
              url: "https://media.giphy.com/media/3xz2BuKH7jQnxBTCtG/giphy.gif"
            }
          },
          liked: true
        }
      ]
    }
  }
  render() {
    return (
      <div className="favorites">
        <h1>Favorites</h1>
        <GifList gifs={this.state.gifs} />
      </div>
    )  
  }
}

export default Favorites;
