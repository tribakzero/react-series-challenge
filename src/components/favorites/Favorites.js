import React, { Component } from 'react';
import { connect } from 'react-redux';
import SearchBar from 'components/searchBar';
import GifList from 'components/gifList';
import { StyledHeading1 } from '../Heading1.style'
import { EmptyPlaceholder } from '../EmptyPlaceholder.style'

export class Favorites extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchValue: ''
    }
  }

  handleSearch = (e) => {
    let value = e.target.value;
    this.setState({
      searchValue: value
    })
  }

  filteredFavorites = () => {
    let result = {};
    let matches = Object.keys(this.props.favorite).filter(favorite => {
      let pattern = new RegExp(`.*?${this.state.searchValue}.*?`, 'i');
      let match = pattern.test(this.props.favorite[favorite].title);
      return match;
    });
    matches.forEach(match => {
      result[match] = { ...this.props.favorite[match] }
    });
    return (result);
  }

  render() {

    var filtered = this.filteredFavorites();
    let searchToggle = null;

    if(Object.keys(this.props.favorite).length === 0) {
      searchToggle = <EmptyPlaceholder>No tienes favoritos.</EmptyPlaceholder>
    } else {
      searchToggle = <GifList gifs={this.props.favorite} />;
      if (this.state.searchValue !== '') {
        if(Object.keys(filtered).length === 0) {
          searchToggle = <EmptyPlaceholder>No hay resultados para esta búsqueda.</EmptyPlaceholder>
        } else {
          searchToggle = <GifList gifs={filtered} />
        }
      }  
    }

    return (
      <React.Fragment>
        <SearchBar value={this.state.searchValue} onChange={this.handleSearch} placeholder="Buscar en favoritos..." />
        <StyledHeading1>Los más chidos... según yo</StyledHeading1>
        { searchToggle }
      </React.Fragment>
    )  
  }
}



const mapStateToProps = (state) => {
  const {
    favorite
  } = state;
  
  return {
    favorite
  };
};

export default connect(mapStateToProps)(Favorites);
