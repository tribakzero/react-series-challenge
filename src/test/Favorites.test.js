import React from 'react';
import { shallow, mount } from 'enzyme';
import { Favorites, handleSearch } from 'components/favorites/Favorites'
import FavoritesConnect from 'components/favorites/Favorites'
import configureStore from 'redux-mock-store'

let wrapper;

beforeEach(() => {
  wrapper = shallow(<Favorites/>)
})

describe('Favorites', () => {
  test('renders properly', () => {
    expect(wrapper).toMatchSnapshot()
  });
  
  test('renders properly with empty favorites', () => {
    wrapper = shallow(<Favorites favorite={{}}/>)
    expect(wrapper).toMatchSnapshot()
  });
  
  test('renders properly with favorites', () => {
    wrapper = shallow(<Favorites favorite={{'id': {}}}/>)
    expect(wrapper).toMatchSnapshot()
  });
  
  test('renders properly with favorites and search', () => {
    wrapper = shallow(<Favorites favorite={{'id': {}}}/>)
    wrapper.setState({
      searchValue: 'TEST'
    });
    expect(wrapper).toMatchSnapshot();
  });
  
  test('renders properly with connect', () => {
    const initialState = {
      favorite: {}
    };
    const mockStore = configureStore();
    let store;
    store = mockStore(initialState)
    let wrapper = shallow(<FavoritesConnect store={store}/>)
    expect(JSON.stringify(wrapper.props().favorite)).toBe(JSON.stringify({}));
  });

  test('handleSearch should call setState', () => {
    wrapper = mount(<Favorites/>)
    wrapper.instance().setState = jest.fn();
    wrapper.instance().handleSearch({target:{value: 'TEST'}})
    expect(wrapper.instance().setState).toHaveBeenCalled();
  })
});
