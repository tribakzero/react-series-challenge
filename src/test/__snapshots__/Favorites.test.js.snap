// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Favorites renders properly with empty favorites 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorite={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <SearchBar
          onChange={[Function]}
          placeholder="Buscar en favoritos..."
          value=""
/>,
        <styled.h1>
          Los más chidos... según yo
</styled.h1>,
        <styled.div>
          No tienes favoritos.
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Buscar en favoritos...",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Los más chidos... según yo",
        },
        "ref": null,
        "rendered": "Los más chidos... según yo",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "No tienes favoritos.",
        },
        "ref": null,
        "rendered": "No tienes favoritos.",
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <SearchBar
            onChange={[Function]}
            placeholder="Buscar en favoritos..."
            value=""
/>,
          <styled.h1>
            Los más chidos... según yo
</styled.h1>,
          <styled.div>
            No tienes favoritos.
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Buscar en favoritos...",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Los más chidos... según yo",
          },
          "ref": null,
          "rendered": "Los más chidos... según yo",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "No tienes favoritos.",
          },
          "ref": null,
          "rendered": "No tienes favoritos.",
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Favorites renders properly with favorites 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorite={
        Object {
            "id": Object {},
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <SearchBar
          onChange={[Function]}
          placeholder="Buscar en favoritos..."
          value=""
/>,
        <styled.h1>
          Los más chidos... según yo
</styled.h1>,
        <Connect(Component)
          gifs={
                    Object {
                              "id": Object {},
                            }
          }
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Buscar en favoritos...",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Los más chidos... según yo",
        },
        "ref": null,
        "rendered": "Los más chidos... según yo",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "gifs": Object {
            "id": Object {},
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <SearchBar
            onChange={[Function]}
            placeholder="Buscar en favoritos..."
            value=""
/>,
          <styled.h1>
            Los más chidos... según yo
</styled.h1>,
          <Connect(Component)
            gifs={
                        Object {
                                    "id": Object {},
                                  }
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Buscar en favoritos...",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Los más chidos... según yo",
          },
          "ref": null,
          "rendered": "Los más chidos... según yo",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "gifs": Object {
              "id": Object {},
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Favorites renders properly with favorites and search 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorite={
        Object {
            "id": Object {},
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <SearchBar
          onChange={[Function]}
          placeholder="Buscar en favoritos..."
          value="TEST"
/>,
        <styled.h1>
          Los más chidos... según yo
</styled.h1>,
        <styled.div>
          No hay resultados para esta búsqueda.
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Buscar en favoritos...",
          "value": "TEST",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Los más chidos... según yo",
        },
        "ref": null,
        "rendered": "Los más chidos... según yo",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "No hay resultados para esta búsqueda.",
        },
        "ref": null,
        "rendered": "No hay resultados para esta búsqueda.",
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <SearchBar
            onChange={[Function]}
            placeholder="Buscar en favoritos..."
            value="TEST"
/>,
          <styled.h1>
            Los más chidos... según yo
</styled.h1>,
          <styled.div>
            No hay resultados para esta búsqueda.
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Buscar en favoritos...",
            "value": "TEST",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Los más chidos... según yo",
          },
          "ref": null,
          "rendered": "Los más chidos... según yo",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "No hay resultados para esta búsqueda.",
          },
          "ref": null,
          "rendered": "No hay resultados para esta búsqueda.",
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
